Added missing geofence options, for Android and some refactoring

vadrian89 opened the pull request:
Hello, took some time because we had a lot of work 

Added the following options to the geofence Zone:
- triggers, used to pass triggers independently
- expirationDuration, used for adding temporary geofences
- dwellLoiteringDelay, used to add a delay for when the device is considered to loiter inside a geofence (Issue #11)
- initialTrigger, provide an initial trigger which is triggered if the conditions are met when adding a geofence

Updated fromJson/toJson functions to include the new fields.

Did some refactoring to try and make the code more readable. Hope it's not an issue with you.

I tried to make more isolated commits, so you can cherry pick what you want and discard what you don't want.

Let me know if you want to modify/revert some of the code, or you find any issues.

byshy commented:
Wow thanks for the contribution, I'll take a look for sure

byshy commented on lib/geofence_foreground_service_method_channel.dart:
```diff
@@ -193,8 +187,7 @@ class MethodChannelGeofenceForegroundService
       areAllAreasRemoved =
-          await foregroundChannel.invokeMethod<bool?>('removeAllGeoFences') ??
-              false;
+          await foregroundChannel.invokeMethod<bool?>('removeAllGeoFences') ?? false;
```
Can we replace the hardcoded values in this file with the new ones created in the ApiMethods class?

vadrian89 replied:
Added ApiMethods class in dart aswell and replaced the method invocations to use the constants.
Also, I've modified the method invocations to directly return the value inside the "try" blocks, to make the code more readable.

byshy commented on android/src/main/kotlin/com/f2fk/geofence_foreground_service/models/Zone.kt:
```diff
@@ -59,28 +57,44 @@ class Zone(
         if (notificationResponsivenessMs != null) {
             result[Constants.notificationResponsivenessMs] = notificationResponsivenessMs
         }
+        // We will convert the even triggers back to a list of integers, just as we receive them
```
Please fix the typo in the comment `even` -> `event`

vadrian89 commented:
Added ApiMethods class in dart aswell and replaced the method invocations to use the constants.
Added _invokeMethod to improve readablity of the code. 
I've modified the method invocations to directly return the value inside the "try" blocks, to make the code more readable.

byshy commented on android/src/main/kotlin/com/f2fk/geofence_foreground_service/GeofenceForegroundServicePlugin.kt:
```diff
@@ -68,134 +67,86 @@ class GeofenceForegroundServicePlugin : FlutterPlugin, MethodCallHandler, Activi
+                iconData = NotificationIconData.fromJson(iconDataJson)
             }
+            serviceIntent = serviceIntent(GeofenceServiceAction.SETUP.toString())
+            val channel = NotificationChannel(
```
In the original code I was checking for the android version O and above, should we bring it back or is it not needed? I remember I was getting a warning to handle it that way

vadrian89 replied:
I wrote lower. The minimum SDK set to the project is above Android O.

byshy commented on android/src/main/kotlin/com/f2fk/geofence_foreground_service/GeofenceForegroundServicePlugin.kt:
```diff
@@ -68,134 +67,86 @@ class GeofenceForegroundServicePlugin : FlutterPlugin, MethodCallHandler, Activi
+            .putExtra(activity!!.extraNameGen(Constants.serviceId), serviceId)
+
+    // Build the pending intent for the intent service
+    private fun servicePendingIntent(intent: Intent) : PendingIntent = PendingIntent.getForegroundService(
```
Also here, in the old code I was checking for the android OS above android O or not (if true then use getForegroundService else getService)

vadrian89 replied:
The minimum SDK is 29 (Android Q), which is above above Android O

byshy commented:
Hi! Apologies for the delayed response—I’ve been juggling quite a bit lately. I took another look and added a couple more notes on a file I hadn’t reviewed last time.

If possible, could you also update the README.md and CHANGELOG.md files with all the recent changes?

Thanks so much for your contribution—I really appreciate it! The remaining notes are super minor.

vadrian89 commented:
Hi, sorry for the late response. 
I was with the holidays when you wrote, then had a lot of work and forgot....
I will answer all your questions.

vadrian89 commented:
Hi again, 
- Updated example and READM to show the fields added to Zone class
- Updated changelog
- Bumped package version to 0.0.14

vadrian89 commented:
Hi, did you manage to look upon the changes?

byshy commented:
Hi, Thanks for the contribution!
unfortunately no, I'm busy this week sadly, I will check it on Thursday-Friday if I have time

Thanks again, I really really appreciate your efforts 🙏
